python/jupyterlab-chat/jupyterlab_chat/ychat.py

```diff
@@ -109,63 +112,52 @@ def get_messages(self) -> list[Message]:
         message_dicts = self._get_messages()
         return [Message(**message_dict) for message_dict in message_dicts]
 
-    def _get_message_by_index(self, index: int):
-        """
-        Return a message from its index.
-        """
-        return self._ymessages[index]
-
     def _get_messages(self) -> list[dict]:
         """
         Returns the messages of the document as dict.
         """
         return self._ymessages.to_py() or []
 
-    def add_message(self, message: Message) -> int:
+    def add_message(self, new_message: NewMessage) -> str:
         """
         Append a message to the document.
         """
         timestamp: float = time.time()
-        message.time = timestamp
+        uid = str(uuid4())
+        message = Message(
+            **asdict(new_message),
+            time=timestamp,
+            id=uid
+        )
+
         with self._ydoc.transaction():
             index = len(self._ymessages) - next((i for i, v in enumerate(self._get_messages()[::-1]) if v["time"] < timestamp), len(self._ymessages))
+            appended = index == len(self._ymessages)
             self._ymessages.insert(
                 index,
                 asdict(message, dict_factory=message_asdict_factory)
             )
-            return index
 
-    def update_message(self, message: Message, index: int, append: bool = False):
+        if appended:
+            self._indexes_by_id[uid] = index
+        else:
+            # if the message has been inserted in the list,
+            self._indexes_by_id = {message["id"]: idx for idx, message in enumerate(self._get_messages())}
```

Comment:
Consider this potential scenario:

- Two frontend Yjs clients (A & B) open an empty chat.
- Both clients write a new message at the same time, but client B has an extra 500ms of latency.
- A's update appends a new item to an empty list, so it is assigned index 0.
- Later, B's update also appends a new item to an empty list, so it is also assigned index 0. Therefore A's index gets bumped from 0 to 1.

I believe this scenario would lead to a bug, where `self._indexes_by_id` would track the index of A's message as `0` instead of `1`.

However, I'm not sure if this scenario would actually happen. I'm OK with this PR being merged with the existing logic, but I think we should also explore if it's more reliable to maintain `self._indexes_by_id` via an observer that listens to every update to the YDoc. That way, it can update the lookup table in response to updates triggered by frontend clients, not just other server extensions calling the Python API.

Let's chat about this more at standup. 👍

Reply:
> it's more reliable to maintain self._indexes_by_id via an observer that listens to every update to the YDoc. That way, it can update the lookup table in response to updates triggered by frontend clients, not just other server extensions calling the Python API.

IIUC this is what does the last commit, by updating the index on any change in the `ymessages`.